access/ms-teams/bot.go

```diff
@@ -125,52 +136,72 @@ func (b Bot) UninstallAppForUser(ctx context.Context, userIDOrEmail string) erro
 	return trace.Wrap(err)
 }
 
-// FetchUser fetches app id for user, installs app for a user if missing, fetches chat id and saves
-// everything to cache. This method is used for priming the cache. Returns trace.NotFound if a
+// FetchRecipient checks if recipient is a user or a channel, installs app for a user if missing, fetches chat id
+// and saves everything to cache. This method is used for priming the cache. Returns trace.NotFound if a
 // user was not found.
-func (b Bot) FetchUser(ctx context.Context, userIDOrEmail string) (*UserData, error) {
+func (b Bot) FetchRecipient(ctx context.Context, recipient string) (*RecipientData, error) {
 	if b.teamsApp == nil {
 		return nil, trace.Errorf("Bot is not configured, run GetTeamsApp first")
 	}
 
 	b.mu.RLock()
-	d, ok := b.users[userIDOrEmail]
+	d, ok := b.recipients[recipient]
 	b.mu.RUnlock()
 	if ok {
 		return &d, nil
 	}
 
-	userID, err := b.getUserID(ctx, userIDOrEmail)
-	if err != nil {
-		return &UserData{}, trace.Wrap(err)
-	}
-
-	var installedApp *msapi.InstalledApp
-
-	installedApp, err = b.graphClient.GetAppForUser(ctx, b.teamsApp, userID)
-	if trace.IsNotFound(err) {
-		err := b.graphClient.InstallAppForUser(ctx, userID, b.teamsApp.ID)
+	// Check if the recipient is a channel
```

Comment:
What happens in the case of 2 simultaneous calls to `FetchRecipient` that miss the cache? I get that the `recipients` map itself is protected by `mu`, but does it matter that you may have multiple copies of the recipient data being generated, and the last write to the cache wins?

Reply:
Maps work by reference (as well as the `*Mutex`)
So, when the map is updated we are actually mutating the map and not some copy.
Being protected by the *Mutex means we are not getting two different states for the map.
I might be missing something, but seems correct to me

But I do agree with you that we should probably change this method to a pointer receiver.

Reply:
@marcoandredinis - It's not the `map` update I was concerned about, that's totally fine. What I was concerned about is was the generation of multiple copies of the `RecipientData`, with them being replaced in the map one after the other. Is it possible someone will have a copy of the stale data? Will it matter? Are there any side effects of that?

And I think @hugoShaka's says "no, it won't" - so with that I'm all good.

Reply:
Oh I'm sorry
Totally missed the point here
We should be good then :+1: